File: Lagrange Function/lagrange_fn.py
# %%
import numpy as np
import sympy as sym

# %%

def lagrange_fn(x_array, y_array, out_type):
    # Inputs-
    # x_array: evaluation points
    # y_array: function values at evaluation points
    #          (associated with the evaluation points of the same index)
    # out_type: 3 options - "fun" returns a function, "sym" returns symbolic
    # math object, "lat" returns latex code for the function

    # for consistency ensure arrays are numpy arrays
    x_array = np.asarray(x_array)
    x_array = np.asarray(x_array)

    # catch easy error
    if len(x_array) != len(y_array):
        raise ValueError('arrays not same length')

    # find length of array
    n = len(x_array)
    # create x
    x = sym.symbols('x')
    # create var
    expr = 0

    for i in range(n):
        # reset term
        term = 1
        for j in range(n):
            if i != j:
                # Add factor
                term = term * ((x - x_array[j]) / (x_array[i] - x_array[j]))

        # addend new term with forrect function value
        expr = expr + y_array[i] * term

    # simplify expression (effect will probably depend on terms)
    # Presumably will convert to the monomial basis(if you aren't a fan you
    # can comment this out)
    expr = sym.simplify(expr)

    # Function case
    if out_type == "fun":
        fun_out = sym.lambdify(x, expr)

    # symbolic math object case
    if out_type == "sym":
        fun_out = expr

    # Latex code case
    if out_type == "lat":
        fun_out = sym.latex(expr)

    return fun_out

    raise ValueError('bad type (need to set out_type as fun, sym, or lat)')

File: Lagrange Function/lagrange_fn.m
function fun_out = lagrange_fn (x_array, y_array, out_type, d)
%Inputs-
%x_array: x-evaluation points
%y_array: function values at evaluation points(associated with the x-value
%of the same index)
%out_type: 3 options - "fun" returns a function, "sym" returns symbolic
%math object, "lat" returns latex code for the function
%d: number of digits desired in printing, only needed for options "sym", "lat"

    %catch easy error
    if length(x_array) ~= length(y_array)
       error('arrays not same length'); 
    end
    
    %find length of array
    n = length(x_array);
    %create x
    x = sym('x');
    %create var
    expr = 0;
    
    for i = 1:n
       %reset term
       term = 1;
       for j = 1:n
           if i ~= j
               %Add factor
               term = term * ((x - x_array(j)) / (x_array(i) - x_array(j))); 
           end
       end
       %addend new term with forrect function value
       expr = expr + y_array(i) * term;
    end
    %simplify expression (effect will probably depend on terms)
    %Presumably will convert to the monomial basis(if you aren't a fan you
    %can comment this out)
    expr = simplify(expr);
    
    %Function case
    if out_type == "fun"
        fun_out = matlabFunction(expr);
        return
    end
    
    %symbolic math object case
    if out_type == "sym"
        %set digits
        digits(d);
        fun_out = simplify(vpa(expr));
        return
    end
    
    
    if out_type == "lat"
        %set digits
        digits(d);
        fun_out = latex(simplify(vpa(expr)));
        return
    end
    
      
    errror('bad type (need to set out_type as fun, sym, or lat)')

    
end
